File: src/lib/llm/prompt.ts
export interface RetrievedSource {
	episodeTitle: string;
	timestamp: string;
	text: string;
}

export const SYSTEM_PROMPT = `You are the Diary of a CEO oracle — a warm, sharp assistant who answers questions using only what guests and Steven Bartlett actually said on the Diary of a CEO podcast.

You will receive numbered transcript excerpts. Ground every claim in them and cite the excerpt number like [1] or [2] after the sentence it supports. Quote short memorable phrases where it helps.

The excerpts rarely answer a question word-for-word — that's expected. When they partially address it, synthesize the closest relevant advice from across the excerpts into a useful answer, and briefly note anything important the excerpts don't cover. Never invent facts that aren't in the excerpts. Only decline to answer when nothing in the excerpts relates to the question at all. Keep answers concise and conversational.

Always answer in the same language as the question, even though the excerpts are in English — translate quoted phrases when needed.`;

/** Assemble the user turn: numbered excerpts followed by the question. */
export function buildGroundedPrompt(question: string, sources: RetrievedSource[]): string {
	const excerpts =
		sources.length === 0
			? '(no relevant excerpts were found in the transcripts)'
			: sources
					.map((s, i) => `[${i + 1}] ${s.episodeTitle} (${s.timestamp})\n${s.text}`)
					.join('\n\n');
	return `Transcript excerpts:\n\n${excerpts}\n\nAnswer in the same language as the question.\n\nQuestion: ${question}`;
}

File: ios/AskDiaryKit/Sources/AskDiaryKit/Prompt.swift
import Foundation

public struct RetrievedSource: Sendable {
    public let episodeTitle: String
    public let timestamp: String
    public let text: String

    public init(episodeTitle: String, timestamp: String, text: String) {
        self.episodeTitle = episodeTitle
        self.timestamp = timestamp
        self.text = text
    }
}

public enum Prompt {
    /// Adapted from the web app (src/lib/llm/prompt.ts) for Apple's on-device
    /// model: it needs a blunter, per-sentence citation rule — the softer web
    /// phrasing made it restructure answers and drop the [n] markers.
    public static let system = """
        You are the Diary of a CEO oracle — a warm, sharp assistant who answers questions using only what guests and Steven Bartlett actually said on the Diary of a CEO podcast.

        You will receive numbered transcript excerpts. Rules, in order of importance:
        1. Every sentence that states a fact MUST end with the number of the excerpt it comes from, in square brackets, like [1] or [3]. An answer without these citation markers is wrong.
        2. Never invent facts that aren't in the excerpts. If the excerpts only partially address the question, synthesize the closest relevant advice and briefly say what they don't cover. Decline only when nothing relates at all.
        3. Write flowing conversational prose, 2–3 short paragraphs, no lists and no headings. Quote short memorable phrases where it helps.

        The user message ends with an instruction naming the language to answer in — follow it, even though the excerpts are in English, and translate quoted phrases when needed.
        """

    /// Assemble the user turn: numbered excerpts, a language instruction, then
    /// the question. Mirrors buildGroundedPrompt in src/lib/llm/prompt.ts.
    public static func grounded(
        question: String, sources: [RetrievedSource], english: Bool = true
    ) -> String {
        let excerpts =
            sources.isEmpty
            ? "(no relevant excerpts were found in the transcripts)"
            : sources.enumerated()
                .map { i, s in "[\(i + 1)] \(s.episodeTitle) (\(s.timestamp))\n\(s.text)" }
                .joined(separator: "\n\n")
        let language = english ? "Answer in English." : "Answer in the same language as the question."
        // the example + reminder sit last on purpose: small models mimic the
        // end of the prompt far more reliably than the system instructions
        return """
            Transcript excerpts:

            \(excerpts)

            \(language)

            Here is the required answer format (note the bracketed excerpt number ending every sentence):
            "Walker says caffeine's half-life keeps it in your system well into the evening [1]. He recommends cutting it off 10 hours before bed [3]."

            Question: \(question)

            Answer in that exact style, but fuller: 2–3 short paragraphs drawing on several excerpts, and every sentence ends with its excerpt number in brackets, like [2].
            """
    }
}
